source/fluid_pressure_boundary_conditions/density.cc

```diff
@@ -0,0 +1,131 @@
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/fluid_pressure_boundary_conditions/density.h>
+#include <aspect/melt.h>
+#include <utility>
+#include <limits>
+
+
+namespace aspect
+{
+  namespace FluidPressureBoundaryConditions
+  {
+
+    template <int dim>
+    void
+    Density<dim>::
+    fluid_pressure_gradient (
+      const types::boundary_id boundary_indicator,
+      const MaterialModel::MaterialModelInputs<dim> &material_model_inputs,
+      const MaterialModel::MaterialModelOutputs<dim> &material_model_outputs,
+      std::vector<Tensor<1,dim> > &fluid_pressure_gradient_outputs
+    ) const
+    {
+      const MaterialModel::MeltOutputs<dim> *melt_outputs = material_model_outputs.template get_additional_output<MaterialModel::MeltOutputs<dim> >();
+      Assert(melt_outputs!=NULL, ExcMessage("Error, MeltOutputs are missing in fluid_pressure_gradient()"));
+      for (unsigned int q=0; q<fluid_pressure_gradient_outputs.size(); ++q)
+        {
+          const Tensor<1,dim> gravity = this->get_gravity_model().gravity_vector(material_model_inputs.position[q]);
+
+          switch (density_formulation)
+            {
+              case DensityFormulation::solid_density:
+              {
+                fluid_pressure_gradient_outputs[q] = material_model_outputs.densities[q] * gravity;
+                break;
+              }
+
+              case DensityFormulation::fluid_density:
+              {
+                fluid_pressure_gradient_outputs[q] = melt_outputs->fluid_densities[q] * gravity;
+                break;
+              }
+
+              default:
+                Assert (false, ExcNotImplemented());
+            }
+        }
+    }
+
+    template <int dim>
+    void
+    Density<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Boundary fluid pressure model");
+      {
+        prm.enter_subsection("Density");
+        {
+          prm.declare_entry ("Density formulation", "solid density",
+                             Patterns::Selection ("solid density|fluid density"),
+                             "The density formulation used to compute the fluid pressure gradient "
+                             "at the model boundary. "
+                             "'solid density' prescribes the gradient of the fluid pressure as "
```

Comment:
Maybe add `"\n\n"` before this line and the one that describes `'fluid density'`, to make it easier to read in the manual.